Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/OptimizedAffinityPropagation.java

```diff
@@ -0,0 +1,318 @@
+          similarities[j][i] = similarity;
+          divergenciesArray[i * (i+1) / 2 + j] = -similarity;
+        }
+      }
+    }
+
+    private double computeMedian(double[] divergenciesArray)
+    {
+
+      double[] doubles = Arrays.stream(divergenciesArray)
+          .flatMap(DoubleStream::of)
+          .distinct()
+          .sorted()
+          .toArray();
+
+      double median;
+      int size = doubles.length;
+      if (size == 0) {
+        median = 0;
+      } else {
+        if (size % 2 == 0)
+          median = (doubles[size / 2] + doubles[size / 2 - 1]) / 2;
+        else
+          median = doubles[size / 2];
+      }
+      return median;
+
+    }
+
+    /**
+     * Updates the responsibilities matrix using the similarity values and the
+     * current availability values.
+     */
+    @Override
+    protected void updateResponsibilities()
+    {
+
+      for(int i = 0 ; i < exampleCount; i++) {
+        double[] rowArray = createSumArray(availabilities[i], similarities[i]);
+        MaxResult maxResult = computeMax(rowArray);
+        for (int j = 0; j < exampleCount; j++) {
+          double maxToSubtract = j != maxResult.maxIndex ? maxResult.maxValue : maxResult.secondMaxValue;
+          responsibilities[i][j] = oneMinusDampingFactor * (similarities[i][j] - maxToSubtract) + dampingFactor * responsibilities[i][j];
+        }
+      }
+    }
+
+    private MaxResult computeMax(double[] rowArray)
+    {
+      MaxResult maxResult = new MaxResult();
+      for (int i = 0; i < rowArray.length; i++) {
+        if(rowArray[i] > maxResult.maxValue){
+          maxResult.secondMaxValue = maxResult.maxValue;
+          maxResult.maxValue = rowArray[i];
+          maxResult.maxIndex = i;
+        }else if(rowArray[i] > maxResult.secondMaxValue){
+          maxResult.secondMaxValue = rowArray[i];
+        }
+      }
+      return maxResult;
+    }
+
+
+    private double[] createSumArray(double[] availability, double[] similarity)
+    {
+      double[] resultArray = new double[availability.length];
+      for (int i = 0; i < availability.length; i++) {
+        resultArray[i] = availability[i] + similarity[i];
+      }
+      return resultArray;
+    }
+
+    /**
+     * Updates the availabilities matrix based on the current responsibility
+     * values.
+     */
+    @Override
+    protected void updateAvailabilities()
+    {
+      ColumnArray rp = new ColumnArray(exampleCount);
+      for (int j = 0; j < exampleCount; j++) {
+        removeNegativeValues(rp, responsibilities, j);
+        rp.array[j] = responsibilities[j][j];
+        rp.sum += responsibilities[j][j];
+        updateAvailabilitiesMatrix(j, rp, availabilities);
+      }
+    }
+
+    private void updateAvailabilitiesMatrix(int j, ColumnArray rp, double[][] availabilities)
+    {
+      double old;
+      for (int i = 0; i < availabilities[0].length; i++) {
+        old = availabilities[i][j];
+        availabilities[i][j] = rp.sum - rp.array[i];

```

Comment:
Minor performance: Not sure if the JIT removes it, but you compute the value, do an array lookup to set it, but then do two subsequent array lookups to use it, then set the array again. Might as well just assign this result to a local variable then use it below.

Reply:
done